sound/soc/sof/ipc4-control.c

```diff
@@ -240,6 +240,50 @@ sof_ipc4_set_generic_control_data(struct snd_sof_dev *sdev,
 	return ret;
 }
 
+static void sof_ipc4_refresh_generic_control(struct snd_sof_control *scontrol)
+{
+	struct sof_ipc4_control_data *cdata = scontrol->ipc_control_data;
+	struct snd_soc_component *scomp = scontrol->scomp;
+	struct sof_ipc4_control_msg_payload *data;
+	struct sof_ipc4_msg *msg = &cdata->msg;
+	size_t data_size;
+	unsigned int i;
+	int ret;
+
+	if (!scontrol->comp_data_dirty)
+		return;
+
+	if (!pm_runtime_active(scomp->dev))
+		return;
+
+	data_size = struct_size(data, chanv, scontrol->num_channels);
+	data = kmalloc(data_size, GFP_KERNEL);
+	if (!data)
+		return;
+
+	data->id = cdata->index;
+	data->num_elems = scontrol->num_channels;
+	msg->data_ptr = data;
+	msg->data_size = data_size;
+
+	scontrol->comp_data_dirty = false;
+	ret = sof_ipc4_set_get_kcontrol_data(scontrol, false, true);
+	msg->data_ptr = NULL;
+	msg->data_size = 0;
+	if (!ret) {
+		for (i = 0; i < scontrol->num_channels; i++) {
+			cdata->chanv[i].channel = data->chanv[i].channel;
+			cdata->chanv[i].value = data->chanv[i].value;
+		}
+	} else {
+		dev_err(scomp->dev, "Failed to read control data for %s\n",
+			scontrol->name);
+		scontrol->comp_data_dirty = true;
```

Comment:
why is this an error? In the previous patch the notification data was tagged as optional, and the commit message also does not hint at any errors "If only a notification without values received, marked the control as dirty
and on next read access fetch the new values from the firmware."

Reply:
This is not the notification handling function, this is a function to read the control data from the firmware with the GET_LARGE_CONFIG message.
Previously we did not needed this as the kernel knew the control's state as it was managing it, but with the notification support we need to be able to read the control's data if the notification did not contained the channel values.

Reply:
not following at all, sorry.

Reply:
When we receive a control change notification it can contain the new values for the control, in that case we will update the kcontrol based on the data received in the notification message.
The the new values for the control is optional, the firmware can send a notification with only the minimal information to identify the control, but no values are included.
In this case we mark the kcontrol as dirty and next time when the user reads the control we will use  GET_LARGE_CONFIG message to read the new values from firmware.
This error print is in this refresh code, it is not the notification handling path.